Modern CMake scripts and MSVC support

btolfa opened the pull request:
---
**By submitting this pull-request, I confirm the following:**

- I have read and understood the contributor guide in [kovri-docs](https://github.com/monero-project/kovri-docs).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.
---

- MSVC 2017 support has been added. Tested on MSVC 2017 with boost 1.66.0
- CMake scripts have been modernized and rewritten [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1)
- The `cpp-netlib` dependency added through `add_subdirectory()` instead of `find_package()` and an external build.

anonimal commented on CMakeLists.txt:
```diff
@@ -267,36 +254,31 @@ if (APPLE OR ANDROID)
   if (WITH_STATIC)
-    #set(CMAKE_FIND_LIBRARY_SUFFIXES .a ${CMAKE_FIND_LIBRARY_SUFFIXES})
-    set(CMAKE_FIND_LIBRARY_SUFFIXES .a)
+    set(OPENSSL_USE_STATIC_LIBS TRUE)
```
OpenSSL is no longer linked statically when building `release-static`. I've only tested on linux and mingw but I imagine this applies to all builds. This line may be the cause or possibly when you decoupled cpp-netlib from `add_subdirectory()`. TBD, but I'll let you find out why.

btolfa replied:
Because cpp-netlib is built via `add_submodule()`, for static linking we need to specify the cmake option `WITH_STATIC_DEPS=ON`. I will fix Makefile accordingly.

btolfa commented on .gitmodules:
```diff
@@ -1,7 +1,7 @@
 	path = deps/cpp-netlib
-	url = https://github.com/anonimal/cpp-netlib.git
-	branch = 0.13-release
+	url = https://github.com/btolfa/cpp-netlib.git
```
First, can you please create a branch which points to anonimal/cpp-netlib@7ab74c50a989225ece5f8d6e2ead83a2938afb50 (the current revision of deps/cpp-netlib) so I can send PR to this branch?

Second, my changes consist of 2 parts - modernize of cmake scripts and C++ code changes required to build under MSVC. In fact, I took the changes for MSVC from the master branch of the cpp-netlib.

I don't like the current way of building and linking with cpp-netlib, so I modernized its build scripts too. 

Adding a dependency to cpp-netlib via `add_subdirectory()` makes build scripts more cross-platform. I also have an idea how to build the cryptopp inside kovri's cmake scripts (through [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html)). Such changes will allow building kovri using only the CMake, without dependency to Makefile (which is difficult to use with MSVC unlike cmake scripts).

anonimal replied:
>First, can you please create a branch which points to anonimal/cpp-netlib@7ab74c5 (the current revision of deps/cpp-netlib) so I can send PR to this branch?

...Why can't you PR to the current revision that's synced upstream? Once that's merged we can bump the submodule.

>In fact, I took the changes for MSVC from the master branch of the cpp-netlib.

That's great, but they should be incorporated into their 0.13-release branch unless otherwise stated.

>Such changes will allow building kovri using only the CMake, without dependency to Makefile

That's great.

btolfa commented on CMakeLists.txt:
The `CMAKE_<LANG>_FLAGS` variable is a bad place for specifying compile definitions - it isn't cross-platform, for example, MSVC has a different compile definition syntax. `add_definitions()` is also inappropriate place because we can't target only release build with it. 

The `set_property()` command (unlike `add_definitions()` supports [generator expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html) and `COMPILE_DEFINITIONS` will generate appropriate command line options all supported compilers.

btolfa commented on CMakeLists.txt:
```diff
@@ -267,36 +254,31 @@ if (APPLE OR ANDROID)
+  set(CPP-NETLIB_BUILD_EXAMPLES OFF CACHE BOOL  "Build the cpp-netlib project examples.") 
+  set(CPP-NETLIB_ENABLE_HTTPS ON CACHE BOOL  "Build cpp-netlib with support for https if OpenSSL is found.") 
+  set(CPP-NETLIB_STATIC_OPENSSL ${WITH_STATIC_DEPS} CACHE BOOL  "Build cpp-netlib using static OpenSSL") 
+  set(CPP-NETLIB_STATIC_BOOST ${WITH_STATIC_DEPS} CACHE BOOL "Build cpp-netlib using static Boost") 
```
I can remove "docstring", but this code is the only way to pass option values to a subproject. 

[Overriding a default option value in CMake from a parent CMakeLists.txt
](https://stackoverflow.com/a/3769269)

btolfa commented on cmake/FindBoost.cmake:
```diff
@@ -0,0 +1,2098 @@
+# file Copyright.txt or https://cmake.org/licensing for details.
+
+#.rst:
+# FindBoost
```
You right, it's really the job of distros. And actually, I need this not for MSVC (where I can install the proper version of CMake), but for FreeBSD, OpenBSD, DragonFly BSD and MinGW builds. Adding this file will solve currently broken [OpenBSD](https://build.getmonero.org/builders/kovri-static-openbsd-amd64) and [DragonFly BSD](https://build.getmonero.org/builders/kovri-static-dragonflybsd-amd64) builds.

Adding this file will allow us to use the latest versions of boost without upgrading cmake.

anonimal replied:
>And actually, I need this not for MSVC (where I can install the proper version of CMake), but for FreeBSD, OpenBSD, DragonFly BSD and MinGW builds. Adding this file will solve currently broken OpenBSD and DragonFly BSD builds.

That's nice that it fixes openbsd and dfly bsd static build, but why do you need it fo freebsd and mingw? We currently have no problems building dynamic/statically on freebsd/mingw platforms.

btolfa replied:
The current cmake scripts throw the following warnings:
FreeBSD buildbot
```
CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:744 (message):
  Imported targets not available for Boost version 106300
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:848 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1435 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:53 (find_package)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:744 (message):
  Imported targets not available for Boost version 106300
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:848 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1435 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:53 (find_package)
```
and
MinGW buildbot
```
CMake Warning at C:/msys64/mingw64/share/cmake-3.10/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  C:/msys64/mingw64/share/cmake-3.10/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  C:/msys64/mingw64/share/cmake-3.10/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:53 (find_package)
```

All these warnings are about `FindBoost.cmake` and we don't have them with upstream version of  `FindBoost.cmake`

anonimal replied:
But warnings are not errors.

If you want to include this file for the sake of resolving backend issues, then we should address the backend issues (if possible) before using this file as a last resort.

btolfa commented on cmake/cotire.cmake:
```diff
@@ -0,0 +1,4190 @@
+# - cotire (compile time reducer)
```
If you don't like having this file in a tree, I can add it as a git submodule from https://github.com/sakra/cotire

Yes, I need it for speeding up MSVC build. On my computer, using of precompiled headers gives a two-fold improvement in build time. Theoretically, the PCH should improve compilation time for other compilers, but in practice, I did not see any significant effects.

PCH does not affect build on Linux, because it is enabled only for MSVC. Most likely, the build time has increased due adding cpp-netlib through `add_subdirectory()`. The subproject inherits all global parent's compile definitions and options. This can be avoided if we add a subproject through [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html) instead of `add_subproject()`.

anonimal replied:
>I can add it as a git submodule from https://github.com/sakra/cotire

Sure.

btolfa commented on src/core/util/filesystem.cc:
```diff
@@ -106,8 +106,8 @@ const std::string StringStream::ReadStringFromByte()
   // Read given amount
-  char buf[len];
-  m_Stream.read(reinterpret_cast<char*>(buf), len);
+  char* buf = reinterpret_cast<char*>(alloca(len));
```
From a pedantic point of [VLA](https://en.wikipedia.org/wiki/Variable-length_array#C99) isn't a feature of C++ but a feature of C99. And as I pointed, MSVC doesn't support it.

Replacing VLA with [alloca()](http://man7.org/linux/man-pages/man3/alloca.3.html) call makes it cross-platform, but in the root, it is still allocating a buffer on the stack.

Anyway, if you don't like it, I can hide it under `#ifdef` directives.

anonimal replied:
>isn't a feature of C++ but a feature of C99

From a pedantic point of view, neither is `alloca`() (a feature of c++).

>Anyway, if you don't like it

It's not about what I like, it's about an unnecessary step that should at least be documented until further refactoring. So, yes, `#ifdef` or leave a TODO.

btolfa commented:
@anonimal - Changes since the last time:
- I've sent PR to anonimal/cpp-netlib#1
I'm going to revert `.gitmodules` changes after PR merge.
- Hidden `filesystem.cc` changes under `#ifdef _MSC_VER`
- Added `cotire` as a git submodule
- Fixed static build (`release-static`, `release-static-deps`)
It works both through Makefile and `WITH_STATIC=ON` and `WITH_STATIC_DEPS=ON` cmake options. It works with GCC, MinGW and MSVC.
- Added support for static runtime linking for MSVC (`/MT` compile flag instead of `/MD`).
According to [cryptopp wiki](https://www.cryptopp.com/wiki/Visual_Studio#Dynamic_Runtime_Linking) static runtime linking has smaller attack surface both on Windows and Linux.
- Add support for building cryptopp inside the kovri cmake project through `ExternalProject_Add()`.
This behavior is disabled by default and can be enabled by `BUILD_CRYPTOPP=ON` cmake option. It is disabled because it is tricky to pass arbitrary options to cryptopp through `ExternalProject_Add()` (like `cryptopp-native` from Makefile). But this option is very convenient for building with MSVC.
This option allows us to build kovri with one cmake command, but this command will be fancy.

About `FindBoost.cmake` - Warnings will turn into errors with modernized cmake scripts because they depend on targets(like `Boost::log`) instead of variables (`Boost_LIBRARIES`).
These errors and warnings can be fixed by updating cmake, or by putting this file into the repository.  
The second solution is simpler.

anonimal commented:
https://github.com/anonimal/cpp-netlib/pull/1 merged. @danrmiller is working on the FindBoost.cmake backend.

btolfa commented:
@anonimal 
- I've replaced changes to `.gitmodules` with cpp-netlib submodule bump btolfa/kovri@b0edf10
- Added fixes for MSVC x64 build and _WIN32_WINNT > 0x0600
- Also opened  anonimal/cpp-netlib#2 with fixes for _WIN32_WINNT > 0x0600

anonimal commented:
anonimal/cpp-netlib#2 merged. Awaiting feedback from Dan about boost backend.

btolfa commented:
@anonimal I can't reproduce the problem with mingw64. On my machine, the build is successful.

```bash
make release-static
```

Can you give the build log from mingw64?

It seems like macOs failure cause by the following line - https://github.com/monero-project/kovri/blob/master/src/CMakeLists.txt#L13, but I didn't change it. Can you give the build log from macOS?

anonimal commented:
>Can you give the build log from mingw64?

[mingw64-release-static.txt](https://github.com/monero-project/kovri/files/1909744/mingw64-release-static.txt) could be related to your `CryptoPP_INCLUDE_DIR` changes in 40d86eb49764bbc36aa393ff3ee1cd178c99f960 but I cannot reproduce the issue on win10 - I can only reproduce on win server 2012 (which is my primary win build env). Both are running cmake 3.10.2.

>It seems like macOs failure cause by the following line - https://github.com/monero-project/kovri/blob/master/src/CMakeLists.txt#L13, but I didn't change it

No, the build is fine with that line, and it's still fine on master. The problem lies elsewhere within this PR. Possibly a CMake caveat or "feature".

anonimal commented on src/client/address_book/impl.cc:
```diff
@@ -89,7 +89,7 @@ void AddressBook::Start(
         m_SharedLocalDestination->GetService());
   m_SubscriberUpdateTimer->expires_from_now(
-      boost::posix_time::minutes(SubscriberTimeout::InitialUpdate));
+      boost::posix_time::minutes{static_cast<std::int64_t>(SubscriberTimeout::InitialUpdate)});
```
and watch out for narrowing conversion warnings. Why do you need this.

anonimal replied:
>I can replace it with the following:

Whatever you do, put an ifdef because MSVC is absolutely absurd. No one should be using microsoft to begin with.

btolfa replied:
It is not an issue with MSVC, it is an issue with Boost 1.67.0
I can revert this change, but the problem with the latest version of Boost will remain.

anonimal replied:
1.67 was released less than 3 days ago, after this thread started. Why were you developing on an unreleased version of boost? 1.67 hasn't even made it to mingw or arch linux yet.

btolfa replied:
1.67 beta is available from March.
>Why were you developing on an unreleased version of boost?
Mainly because I can. As long as it does not break the build with older versions of Boost, using the newer version is not an issue.

> 1.67 hasn't even made it to mingw or arch linux yet.
It does not stop users from building it from source. And for some distros (like Ubuntu 14.04 and 16.04) it is the only possible solution to receive the recent version of Boost.

anonimal replied:
>Mainly because I can. 

That's great. Good for you. If you want to PR to this project, you need to work *with* our build system - not against it.

>As long as it does not break the build with older versions of Boost, using the newer version is not an issue.

Adding superfluous code and overhead is an issue.

>It does not stop users from building it from source. And for some distros (like Ubuntu 14.04 and 16.04) it is the only possible solution to receive the recent version of Boost.

See note about collaboration.

btolfa commented:
@anonimal it looks like the issue with mingw64-release build is related to https://github.com/weidai11/cryptopp/issues/442

Can you please make clean `git clone` or run `git submodule update` before build?
It seems like submodule `deps/cryptopp` on that machine is outdated.

anonimal commented:
>@anonimal it looks like the issue with mingw64-release build is related to weidai11/cryptopp#442

No, that has been long since been resolved. This is against a fresh clone. Please fix your new cmake to work on all available platforms.

anonimal commented:
@btolfa Once you fix the breakages and other issues, and rebase, this PR can be merged.  As for having findboost in this tree, since https://github.com/monero-project/meta/issues/201 isn't going anywhere, we can keep that in this PR and open a new ticket to remove it once https://github.com/monero-project/meta/issues/201 is resolved.

btolfa commented:
@anonimal 
- I have moved fixes for Boost 1.67 into a separate PR #855 #856 
- I have fixed osx release-static build
- I still can't reproduce issue with mingw release-static build

btolfa commented:
I've fixed these issues.
> make release-static on Ubuntu, CMake 3.5.1, gcc 5.4.0:

I can't reproduce this issue.
> make release-static on macOS 10.13:
